File: src/core/qpair.rs
use std::hash::Hash;

use cxx::{ExternType, type_id};
use cxx_qt_lib::QByteArray;

/// Typedef for `std::pair<T1, T1>`.
#[repr(C)]
#[derive(Clone, Debug, Default, PartialEq, Eq, PartialOrd, Ord, Hash)]
pub struct QPair<T1, T2> {
    pub first: T1,
    pub second: T2,
}

impl<T1, T2> From<(T1, T2)> for QPair<T1, T2> {
    fn from((first, second): (T1, T2)) -> Self {
        Self { first, second }
    }
}

impl<T1, T2> From<QPair<T1, T2>> for (T1, T2) {
    fn from(value: QPair<T1, T2>) -> Self {
        (value.first, value.second)
    }
}

impl<'a, T1, T2> From<&'a QPair<T1, T2>> for (&'a T1, &'a T2) {
    fn from(value: &'a QPair<T1, T2>) -> Self {
        (&value.first, &value.second)
    }
}

macro_rules! impl_extern {
    ( $firstTypeName:ty, $secondTypeName:ty, $typeId:literal ) => {
        // SAFETY: Static checks on the C++ side to ensure the size is the same.
        unsafe impl ExternType for QPair<$firstTypeName, $secondTypeName> {
            type Id = type_id!($typeId);
            type Kind = cxx::kind::Trivial;
        }
    };
}

impl_extern!(QByteArray, QByteArray, "QPair_QByteArray_QByteArray");

#[cfg(feature = "net")]
impl_extern!(crate::QHostAddress, i32, "QPair_QHostAddress_i32");

File: src/core/qpair.cpp
#include "cxx-qt-io/qpair.h"

#include <cxx-qt-io/assertion_utils.h>

#define CXX_QT_IO_QPAIR_ASSERTS(firstTypeName, secondTypeName, combinedName)   \
  assert_alignment_and_size(QPair_##combinedName, {                            \
    firstTypeName first;                                                       \
    secondTypeName second;                                                     \
  });                                                                          \
                                                                               \
  static_assert(QTypeInfo<QPair_##combinedName>::isRelocatable);               \
                                                                               \
  static_assert(::std::is_copy_assignable<firstTypeName>::value);              \
  static_assert(::std::is_copy_constructible<firstTypeName>::value);           \
  static_assert(::std::is_move_constructible<firstTypeName>::value);           \
  static_assert(::std::is_copy_assignable<secondTypeName>::value);             \
  static_assert(::std::is_copy_constructible<secondTypeName>::value);          \
  static_assert(::std::is_move_constructible<secondTypeName>::value);

CXX_QT_IO_QPAIR_ASSERTS(QByteArray, QByteArray, QByteArray_QByteArray);

#ifdef CXX_QT_IO_NET_FEATURE
CXX_QT_IO_QPAIR_ASSERTS(QHostAddress, ::std::int32_t, QHostAddress_i32);
#endif
